arch/arm/include/aarch32/cortex_m/wrap_func.h

```diff
@@ -0,0 +1,131 @@
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ * @brief
+ */
+
+#ifndef ZEPHYR_ARCH_ARM_INCLUDE_AARCH32_CORTEX_M_WRAP_FUNC_H_
+#define ZEPHYR_ARCH_ARM_INCLUDE_AARCH32_CORTEX_M_WRAP_FUNC_H_
+
+#ifdef _ASMLANGUAGE
+
+/* nothing */
+
+#else
+
+/**
+ * @brief Macro for "sandwiching" a function call (@p name) in two other calls
+ *
+ * This macro should be called via @ref WRAP_FUNC or @ref WRAP_FUNC_STACK_ARGS.
+ *
+ * This macro creates the function body of an "outer" function which behaves
+ * exactly like the wrapped function (@p name), except that the preface function
+ * is called before, and the postface function afterwards.
+ *
+ * @param preface   The function to call first. Must have no parameters and no
+ *                  return value.
+ * @param name      The main function, i.e. the function to wrap. This function
+ *                  will receive the arguments, and its return value will be
+ *                  returned.
+ * @param postface  The function to call last. Must have no parameters and no
+ *                  return value.
+ * @param store_lr  The assembly instruction for storing away the LR value
+ *                  before the functions are called.
+ * @param load_lr   The assembly instruction for restoring the LR value after
+ *                  the functions have been called.
+ */
+#define WRAP_FUNC_RAW(preface, name, postface, store_lr, load_lr) \
+	do { \
+		__asm(".global "#preface"; .type "#preface", %function"); \
+		__asm(".global "#name"; .type "#name", %function"); \
+		__asm(".global "#postface"; .type "#postface", %function"); \
+		\
+		__asm(store_lr); \
+		__asm("push {r0-r3}"); \
+		__asm("bl "#preface); \
+		__asm("pop {r0-r3}"); \
+		__asm("bl "#name); \
+		__asm("push {r0-r3}"); \
+		__asm("bl "#postface); \
+		__asm("pop {r0-r3}"); \
+		__asm(load_lr); \
+	} while (0)
+
+/**
+ * @brief Macro for "sandwiching" a function call (@p name) in two other calls
+ *
+ * This is for functions which do not pass arguments or return values via the
+ * stack. I.e. the arguments and return values must each fit within 4 words,
+ * after accounting for alignment.
+ *
+ * If nothing is passed on the stack, the stack can be used to store LR without
+ * messing with offsets.
+ *
+ *	int foo(char *arg); // Implemented elsewhere.
+ *	int __attribute__((naked)) foo_wrapped(char *arg)
+ *		{WRAP_FUNC(bar, foo, baz)}
+ *
+ * is equivalent to
+ *
+ *	int foo(char *arg); // Implemented elsewhere.
+ *	int foo_wrapped(char *arg)
+ *	{
+ *		bar();
+ *		int res = foo(arg);
+ *		baz();
+ *		return res;
+ *	}
+ *
+ * @note __attribute__((naked)) is not mandatory, but without it, GCC gives a
+ *       warning for functions with a return value. It also saves a bit of space
+ *       since it removes a little code that is not necessary.
+ *
+ * See @ref WRAP_FUNC_RAW for more information.
+ */
+#define WRAP_FUNC(preface, name, postface) \
+	WRAP_FUNC_RAW(preface, name, postface, "push {lr}", "pop {pc}")
```

Comment:
This needs to satisfy the stack alignment, so you need to push an even number of registers

Reply:
Also, this should be made compatible to Baseline Cortex-M

Reply:
> This needs to satisfy the stack alignment, so you need to push an even number of registers

So, I just push a dummy register together with it? Maybe r4?

> Also, this should be made compatible to Baseline Cortex-M

What does this mean?

Reply:
You can see in fault.S or swap_helper.S how push/pop is different for Baseline and Mainline.
Yeap, a dummy register would probably do.

> What does this mean?

I think in Baseline you cannot pop LR directly. But might be more restrictions, so pls, take care of them :)

Reply:
I cannot find any documentation describing the differences between baseline and mainline. I'm popping to PC, is that different from popping to LR?

Reply:
Yeap, PC is R15, LR is R14.
See in the v6-M manual: https://static.docs.arm.com/ddi0419/e/DDI0419E_armv6m_arm.pdf

Reply:
BTW looks like popping to PC is ok in all variants.

    _oops:
        push {r0, lr}
        bl z_do_kernel_oops
        /* return from SVC exception is done here */
        pop {r0, pc}

Reply:
To clarify: I did not spot any errors other than the stack alignment, but you should check you make this ASM compatible with v6-M architecture.

Reply:
Should be good now for baseline (tried on nrf51). I only need to fix the SP test.

Reply:
once you fix CI you can bring this out of draft, and set the milestone